chapters/packages.tex

```diff
@@ -321,3 +321,132 @@ \section{External Resources}\label{external-resources}
+The pattern \filename{$\mathit{language}$} stands for the ISO 639-1 language code, e.g., \filename{de} or \filename{en}.
+\end{enumerate}
+
+The detailed format of these files is described in \href{https://www.gnu.org/software/gettext/manual/gettext.pdf}{GNU gettext}.
+For Modelica translations, only the keywords \lstinline!msgctxt!, \lstinline!msgid! and \lstinline!msgstr! are used, meaning that a translation entry looks like this:
+\begin{lstlisting}
+#: $\mathit{filename}$:$\mathit{lineNumber}$
+msgctxt $\mathit{messageContext}$
+msgid $\mathit{messageIdentifier}$
+msgstr $\mathit{messageTranslation}$
+\end{lstlisting}
+
+\begin{nonnormative}
+In the remainder of this section, several facts about the gettext specification are interleaved non-normatively for easy access to some of the gettext basics.
+Always refer to the external gettext specification for full detail or in case of doubt.
+
+All text strings are in double quotes and encoded with UTF-8 characters.
+Comments start with an \lstinline!#! and are continued until the end of line.
+Spaces outside strings are ignored and used as separators.
+
+The files consist of a header and a body.
+The header is marked with an empty \lstinline!msgid! and looks like this:
+\begin{lstlisting}
+# SOME DESCRIPTIVE TITLE.
+# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
+# This file is distributed under the same license as the PACKAGE package.
+# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
+#
+#, fuzzy
+msgid ""
+msgstr ""
+"Project-Id-Version: PACKAGE VERSION\n"
+"Report-Msgid-Bugs-To: \n"
+"POT-Creation-Date: 2019-03-15 10:52+0100\n"
+"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
+"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
+"Language-Team: LANGUAGE <LL@li.org>\n"
+"Language: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+\end{lstlisting}
+All general terms in the header should be replaced by specific information.
+
+Following the header, there is one translation entry for each string to be translated.
+It can start with an optional comment describing the location (file name and line number) of the text to translate.
+Multiple occurences of the same string can be listed here, separated by space.
+\end{nonnormative}
+
+The $\mathit{messageContext}$ following the keyword \lstinline!msgctxt! shall be the full name of the Modelica class (e.g., \lstinline!"Modelica.Blocks.Math.Sin"!) where the text appears.
+Short class definitions do not appear here.
+Texts in such classes belong to the enclosing full class definition.
+
+The text string which shall be translated is used as $\mathit{messageIdentifier}$ (following the \lstinline!msgid! keyword), and shall contain the original string from the Modelica code.
+% Describe down-quoting and concatenation of Modelica string literals.
+Note that if a \lstinline!msgid! string is given more than once in the same context, all occurrences are translated with the same (last) translation!
+
```

Comment:
Maybe an example is needed to clarify:
```suggestion

\begin{example}
Suppose the following expression appears in a context to be translated:
\begin{lstlisting}
  "Frequency: " + String(2 * Modelica.Constants.pi / T) + "rad/s\n"
+ "Period: %{T}"
\end{lstlisting}
The two strings to be used as $\mathit{messageIdentifier}$ are \lstinline!"Frequency: "! and \lstinline!"rad/s\nPeriod: %{T}"!.
\end{example}

```

Reply:
Didn't we say that we only want to concatenate description strings? In such strings there are only string literals and no references to other variables.

Reply:
I'm afraid that was a misunderstanding.  It is true that the situation is particularly clear for _description-string_ (which is `[ STRING { "+" STRING } ]` in the grammar), but the possibility to break long string literals apart is present and frequently used also elsewhere in Modelica source code.

To be it seems beyond the power of this MCP to mandate that the use habit of breaking up long literals is abandoned in order to get sensible text units to translate.

Maybe the example would be even better if there was also a `String` constant (for example `ModelicaServices.target`) being concatenated, to clarify that `String` constants shall not be treated similar to literals when it comes to translation?

Reply:
It was never the goal of this MCP to support the translation of strings in general expressions. because this would change probably the simulation result, or even lead to errors during simulation, if the strings are used to control something.
To enable translation of strings in expressions the modeler has to mark them as subject for translation, and then a tool can take care of it. But this would be another MCP.
The translation of descriptive strings as a first step is a significant improval for users which are not so familiar with the english language, and the missing translation of strings in expression should not hinder the completion of this MCP.

Reply:
Yes applying translations anywhere in a model does indeed seem dangerous, but fortunately the MCP clearly defines the context where strings are to be translated, so there should be no risk of altering any simulation results.

Reply:
I agree with Gerd here. It should only be description-strings (that must be literal or concatenated literals) and info-strings - and we should also restrict the latter to literal or concatenated literals.

Reply:
But wouldn't that require that this MCP fully restricts `Documentation.info` and `Documentation.revisions` according to the same grammar as a description string? The idea isn't totally strange to me, but I still want to make sure this is really the direction we want to take this.

Why would one stop at `Documentation.info` and `Documentation.revisions`? Can you give a single example of one of the listed annotations that it would not make sense to restrict in the same way?

Reply:
2: Yes, it should apply to all of them.
1: For this MCP it suffices that only fully literal literal strings are supported for translation.

Reply:
I don't agree that just single string literals is good enough for this MCP. A solution that only allows translation of the annotations that happen to be given without the use of literal concatenation just wouldn't meet my expectations of a ready MCP.

Note that I don't consider it an alternative to say that annotations given by concatenated literals should be translated piece by piece; this would just result in a mess for translators.

Reply:
I meant string literals and concatenated string literals; so as soon as there is a non-literal string the entire text is ignored.

Reply:
While that approach is safe in some sense, it seems we're only half done if we define a mechanism for translation without making sure that it is applicable to real-world libraries out there.  I see a chance to really strengthen this MCP by making sure that the things we would like to cover with translation really are available to translation.

To support this discussion, I think we should ask someone like @qlambert-pro to make a prototype for POT file creation that will tell us about any annotations that we say we'd like to translate, but that don't meet the concatenated string literal requirement.

Reply:
You mean another prototype beside the SimulationX specific one from https://github.com/ESI-ITI-GmbH/TranslationTest/blob/master/Scripts/SimulationX/Generate-Template.py ?

Reply:
By the way @beutlich, it's nice to see you here in the discussions of this PR!

Reply:
I agree that an example would be helpful, but then an example without those problems.